Source/DataSources/KmlDataSource.js

```diff
@@ -1846,6 +2384,210 @@ define([
+                    networkLink.updating = false;
+                    networkLinks.splice(networkLinks.indexOf(networkLink), 1);
+                    return;
+                }
+                networkLink.cookie = defaultValue(queryStringValue(networkLinkControl, 'cookie', namespaces.kml), '');
+                minRefreshPeriod = defaultValue(queryNumericValue(networkLinkControl, 'minRefreshPeriod', namespaces.kml), 0);
+            }
+
+            var now = JulianDate.now();
+            var refreshMode = networkLink.refreshMode;
+            if (refreshMode === RefreshMode.INTERVAL) {
+                if (defined(networkLinkControl)) {
+                    networkLink.time = Math.max(minRefreshPeriod, networkLink.time);
+                }
+            } else if (refreshMode === RefreshMode.EXPIRE) {
+                var expires;
+                if (defined(networkLinkControl)) {
+                    expires = queryStringValue(networkLinkControl, 'expires', namespaces.kml);
+                }
+                if (defined(expires)) {
+                    try {
+                        var date = JulianDate.fromIso8601(expires);
+                        var diff = JulianDate.secondsDifference(date, now);
+                        if (diff > 0 && diff < minRefreshPeriod) {
+                            JulianDate.addSeconds(now, minRefreshPeriod, date);
+                        }
+                        networkLink.time = date;
+                    } catch (e) {
+                        console.log('KML - NetworkLinkControl expires is not a valid date');
+                        remove = true;
+                    }
+                } else {
+                    console.log('KML - refreshMode of onExpire requires the NetworkLinkControl to have an expires element');
+                    remove = true;
+                }
+            }
+
+            var networkLinkEntity = networkLink.entity;
+            var entityCollection = dataSource._entityCollection;
+            var newEntities = newEntityCollection.values;
+
+            function removeChildren(entity) {
+                entityCollection.remove(entity);
+                var children = entity._children;
+                var count = children.length;
+                for(var i=0;i<count;++i) {
+                    removeChildren(children[i]);
+                }
+            }
+
+            // Remove old entities
+            var entitiesCopy = entityCollection.values.slice();
+            for (var i=0;i<entitiesCopy.length;++i) {
+                if (entitiesCopy[i].parent === networkLinkEntity) {
+                    entitiesCopy[i].parent = undefined;
+                    removeChildren(entitiesCopy[i]);
+                }
+            }
+
+            // Add new entities
+            entityCollection.suspendEvents();
+            for (i = 0; i < newEntities.length; i++) {
+                if (!defined(newEntities[i].parent)) {
```

Comment:
I won't mention it again, but anywhere you access a property or look up an index more than once, always extract out a local variable.  JavaScript is slow and dumb so we have to perform these manual optimizations.